### PR TITLE
Refactor: add e2e test using ginkgo framework

### PR DESCRIPTION
**Pull request by ycyaoxdu**
Signed-off-by: ycyaoxdu <yaoyuchen0626@163.com>

opened for issue: https://github.com/open-cluster-management-io/community/issues/118 , have refactored the e2e(functional) test using ginkgo framework.

### CODE REVIEW

File: test/functional/clusteradm/clusteradmE2E_test.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+	"testing"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+
+	"open-cluster-management.io/clusteradm/test/functional/util"
+)
+
+var e2e, e2e2, e2e3 *util.TestE2eConfig
```
Comment (zhiweiyin318) [inline test/functional/clusteradm/clusteradmE2E_test.go:16]: it is better to define the var in each case

File: test/functional/clusteradm/clusteradmE2E_test.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		// clean
+		e2e2.KindDeleteCluster()
+	})
+
+	ginkgo.Context("join hub scenario", func() {
+		ginkgo.By("join hub with bootstrap token")
+
+		ginkgo.It("should managedcluster join hub successfully", func() {
+			ginkgo.By("clusteradm version check")
+			err := e2e2.RunCmd(e2e2.Streams, "version")
```
Comment (zhiweiyin318) [inline test/functional/clusteradm/clusteradmE2E_test.go:139]: e2e2.RunCmd("version")
Streams is the filed of e2e2, not need as input parameter.
use the name clusteradm("version")  may be better understand.

File: test/functional/clusteradm/clusteradmE2E_test.go (hunk 1 above)
Comment (ycyaoxdu) [reply test/functional/clusteradm/clusteradmE2E_test.go:16]: ok

File: test/functional/clusteradm/clusteradmE2E_test.go (hunk 2 above)
Comment (ycyaoxdu) [reply test/functional/clusteradm/clusteradmE2E_test.go:139]: The output will be redirected in some case, such as this: :https://github.com/open-cluster-management-io/clusteradm/pull/111/files#diff-d092bd45baec86399bbd7cf1c6a927c365a7db10ce559b3aca73e35604fce025R73 .
So i added the stream parameter. 
And I'm wondering whether i should remove this stream parameter and then define another function to handle these cases, or keep this parameter as it is?

And for the function name, i think your suggestion is really good, i will rename that function.

**Comment by ycyaoxdu** [comment]
#84 occurs

File: test/functional/clusteradm/clusteradmE2E_test.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+	"open-cluster-management.io/clusteradm/test/functional/util"
+)
+
+func TestE2EClusteradm(t *testing.T) {
+	gomega.RegisterFailHandler(ginkgo.Fail)
+	ginkgo.RunSpecs(t, "E2E clusteradm test")
+}
+
+var _ = ginkgo.Describe("test clusteradm with service account", func() {
```
Comment (zhiweiyin318) [inline test/functional/clusteradm/clusteradmE2E_test.go:18]: suggest  use different files to include different cases.
you can refer https://github.com/open-cluster-management-io/registration-operator/tree/main/test/e2e
usually we have a e2e entry file named [e2e_suite_test.go](https://github.com/open-cluster-management-io/registration-operator/blob/main/test/e2e/e2e_suite_test.go)

File: test/functional/clusteradm/clusteradmE2E_test.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+		ginkgo.By("join hub with service account")
+
+		ginkgo.It("should managedcluster join hub successfully", func() {
+			ginkgo.By("clusteradm version check")
+			err := e2e.Clusteradm().Version().Run()
+			gomega.Expect(err).NotTo(gomega.HaveOccurred(), "clusteradm version check error")
+
+			ginkgo.By("init hub")
+			jn, err := e2e.Clusteradm().Init(
+				"--context", e2e.Cluster().Hub().Context(),
```
Comment (zhiweiyin318) [inline test/functional/clusteradm/clusteradmE2E_test.go:48]: the kind clusters are created before run e2e, right? 
could we set the info of clusters as a variables, make it easy to understand, or add more comments?

File: test/e2e/clusteradm/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+---
+## Cluster()
+You can get cluster informations by call Cluster(). Such as:
+```
+// name of hub
+e2e.Cluster().Hub().Name()
+
+// context of managedcluster1
+e2e.Cluster().ManagedCluster1().Context()
```
Comment (qiujian16) [inline test/e2e/clusteradm/README.md:24]: Is this based on the assumption that all tests are run on kind? Is it possible that we have another kuberentes (non-kind) to run e2e. It should be also doc'ed here.

Comment (ycyaoxdu) [reply test/e2e/clusteradm/README.md:24]: yes, assuming that we use kind to start clusters, i will add comments to explain it.

**Comment by ycyaoxdu** [comment]
add non-kind clusters support, user needs to specify the cluster create command format before setup, https://github.com/open-cluster-management-io/clusteradm/pull/111/files#diff-8a2c59005a01c6a3ffdca09c00a682adbebf7800f6bb6551391260815569133fR18-R23 
@qiujian16 now user can determine what kubernetes cluster they use.

**Comment by ycyaoxdu** [comment]
RESTConfig  seems redundant, removed.

**Comment by ycyaoxdu** [comment]
rebased

File: test/e2e/clusteradm/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+`Clean(args ...string)` ,
+`Install(args ...string)` ,
+`Proxy(args ...string)` ,
+`Unjoin(args ...string)`.
+
+---
+### reset e2e environment
+
+#### 1. ResetEnv
+First, we define an intial e2e environment state: 1 hub, 2 managed(cluster1, cluster2), hub is intialized, cluster1 is join and accepted, cluster2 is just created.
```
Comment (qiujian16) [inline test/e2e/clusteradm/README.md:55]: initial

File: test/e2e/clusteradm/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+---
+
+
+## write test
+
+Now we will show you how to write tests for clusteradm command.
+
+0.  Before we start, you should figure out what cluseter are you using, and the corresponding `create cluster` command format, `delete cluster` command format and `cluster context` format, which you should use `<cluster-name>` to replace the real cluster name. See the following example:
+
+    If we're using [kind](), the create cluster command format would be `kind crea  te cluster --name <cluster-name>`, the delete cluster command format would be `kind delete cluster --name <cluster-name>`, the context format would be `kind-<cluster-name>`.
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply test/e2e/clusteradm/README.md:77]: is [kind]() correct?

**Review by qiujian16 — COMMENTED** [review]
thanks, I think it makes more sense now, Just some nit. It looks good in general

/approve

File: test/e2e/clusteradm/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ycyaoxdu) [reply test/e2e/clusteradm/README.md]: actually, this part is departed, sorry for forget to modify this paragraph. 
 and i have rewrote the readme

**Comment by ycyaoxdu** [comment]
/hold 
@qiujian16 
I think `cmd.run()` is still not easy to understand. 
So I'm trying to remove `run()/output()` method and save the `host` and `token` in a struct member, rather than a return value, then we can use these values by calling a method looks like `e2e.xxxx.Host()`.

**Comment by ycyaoxdu** [comment]
> @qiujian16 I think `cmd.run()` is still not easy to understand. So I'm trying to remove `run()/output()` method and save the `host` and `token` in a struct member, rather than a return value, then we can use these values by calling a method looks like `e2e.xxxx.Host()`.

/cc @qiujian16  done. 
1. Changed the command interface. 
2. migrate test framework from `ginkgo/v1` to `ginkgo/v2` (only for e2e test).
Please help to review and give more suggestions. thanks.

**Pull request merged**